### PR TITLE
Add quizzes to DFS and Flood Fill Units

### PR DESCRIPTION
**Pull request by xyzyzl**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

**Comment by xyzyzl** [comment]
still tentative

File: content/3_Silver/Flood_Fill.mdx [hunk 1]
```diff
@@ -739,3 +739,23 @@ print(room_num)
 ## Problems
 
 <Problems problems="general" hideSuggestProblemButton />
+
+## Quiz
+
+<Quiz>
+	<Quiz.Question>
+		What would be a valid function header for DFS on an undirected graph to work *without* any other variables?
```
Comment (bqi343) [inline content/3_Silver/Flood_Fill.mdx:747]: none of these? don't you need to keep track of what's visited and what's not?

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/3_Silver/Flood_Fill.mdx (hunk 1 above)
Comment (xyzyzl) [reply content/3_Silver/Flood_Fill.mdx:747]: meant to say tree

**Comment by dutinmeow** [comment]
Where's the rest?

File: content/3_Silver/Flood_Fill.mdx [hunk 2]
```diff
@@ -739,3 +739,23 @@ print(room_num)
...
+		<Quiz.Answer correct>
```
Comment (dutinmeow) [inline content/3_Silver/Flood_Fill.mdx:748]: Wouldn't both the first and second answer work? The second one is just much slower

**Review by dutinmeow — CHANGES_REQUESTED** [review]

File: content/3_Silver/Flood_Fill.mdx (hunk 2 above)
Comment (xyzyzl) [reply content/3_Silver/Flood_Fill.mdx:748]: doesn't it get stuck in a dead loop unless you have an external visited array?

**Comment by xyzyzl** [comment]
> 
> 
> Where's the rest?

haven't done yet, i just need to check if i formatted it correctly. i'll update this PR over the next few days with the remaining problems

File: content/3_Silver/Flood_Fill.mdx (hunk 2 above)
Comment (xyzyzl) [reply content/3_Silver/Flood_Fill.mdx:748]: nevermind you're right

**Comment by xyzyzl** [comment]
i don't think markdown reacts well to codeblocks in the thing, i will make another

File: content/3_Silver/Flood_Fill.mdx [hunk 3]
```diff
@@ -739,3 +739,66 @@ print(room_num)
...
+
+## Quiz
+
+<Quiz>
+	<Quiz.Question>
+		How many times will we have to run DFS to perform a flood fill on a graph G with 4 connected components?
+		<Quiz.Answer>
+			1
+			<Quiz.Explanation>
+				Incorrect. Each DFS can travel one connected component, so a graph with four connected components would require **4** DFS runs.
```
Comment (Aadit-Ambadkar) [inline content/3_Silver/Flood_Fill.mdx:751]:
```suggestion
				Incorrect. Each DFS can travel one connected component, so a graph with four connected components would require $\mathbf{4}$ DFS runs.
```

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: content/3_Silver/Flood_Fill.mdx (hunk 2 above)
Comment (xyzyzl) [reply content/3_Silver/Flood_Fill.mdx]: ok so i specifically said O(N) now

File: content/3_Silver/DFS.mdx [hunk 4]
```diff
@@ -585,3 +585,77 @@ public class BuildingTeams
 ### Problems
 
 <Problems problems="bip" hideSuggestProblemButton />
+
+## Quiz
+<Quiz>
+	<Quiz.Question>
+		What would be a valid function header for DFS on an undirected tree to work in linear time *without* any other variables?
+		<Quiz.Answer correct>
+			`void dfs(int v, int p, vector<vector<int> >& adj)`
```
Comment (dutinmeow) [inline content/3_Silver/DFS.mdx:594]:
```suggestion
			`void dfs(int v, int p, vector<vector<int>>& adj)`
```

I don't think these spaces are necessary post C++98?

File: content/3_Silver/DFS.mdx [hunk 5]
```diff
@@ -585,3 +585,77 @@ public class BuildingTeams
...
+			<Quiz.Explanation>
+				Incorrect. This DFS function cannot track the current node, therefore it cannot even run.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+	</Quiz.Question>
+	
+	<Quiz.Question>
+		How is a preorder traversal found for a binary tree?
+		<Quiz.Answer correct>
+			Add the current node to the list, and then run DFS on all child nodes of the current node.
```
Comment (dutinmeow) [inline content/3_Silver/DFS.mdx:624]: Specify the order in which you visit the child nodes

File: content/3_Silver/DFS.mdx [hunk 6]
```diff
@@ -585,3 +585,77 @@ public class BuildingTeams
...
+		<Quiz.Answer>
+			Run DFS on all child nodes of the current node, then add the current node to the list.
+			<Quiz.Explanation>
+				Incorrect. This is postorder traversal.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+	</Quiz.Question>
+
+	<Quiz.Question>
+		If a graph has two connected components, can DFS run through all nodes at once?
```
Comment (dutinmeow) [inline content/3_Silver/DFS.mdx:650]:
```suggestion
		If a graph has two connected components, does there exist a node where we could start a DFS that visits every node in the graph?
```
It's unclear what you mean by DFS (I think by default, DFS generates a spanning forest)

Comment (dutinmeow) [inline content/3_Silver/DFS.mdx:650]: I would phrase this as a true/false question

File: content/3_Silver/Flood_Fill.mdx [hunk 7]
```diff
@@ -739,3 +739,66 @@ print(room_num)
 ## Problems
 
 <Problems problems="general" hideSuggestProblemButton />
+
+## Quiz
+
+<Quiz>
+	<Quiz.Question>
+		How many times will we have to run DFS to perform a flood fill on a graph G with 4 connected components?
```
Comment (dutinmeow) [inline content/3_Silver/Flood_Fill.mdx:747]: Again, this is unclear
```suggestion
		What is the minimum number of starting nodes for flood fills we would need to visit every node in an undirected graph with four components?
```

File: content/3_Silver/Flood_Fill.mdx [hunk 8]
```diff
@@ -739,3 +739,66 @@ print(room_num)
...
+		<Quiz.Answer>
+			The number of nodes in each connected component
+			<Quiz.Explanation>
+				Incorrect. We can find the number of nodes in each connected component by iterating through all nodes after running floodfill and determining which connected component each node is in.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+		<Quiz.Answer correct>
+			The shortest path between two nodes
+			<Quiz.Explanation>
+				Correct! This cannot be found with floodfill, because floodfill doesn't keep track of distances.
```
Comment (dutinmeow) [inline content/3_Silver/Flood_Fill.mdx:791]: not sure how I feel about this :|

Comment (bqi343) [reply content/3_Silver/Flood_Fill.mdx:791]: I think this question is okay. You can use BFS to flood fill and maintain shortest distances at the same time, but we haven't described how to maintain shortest distances in this module.

File: content/3_Silver/DFS.mdx [hunk 9]
```diff
@@ -585,3 +585,77 @@ public class BuildingTeams
...
+## Quiz
+<Quiz>
+	<Quiz.Question>
+		What would be a valid function header for DFS on an undirected tree to work in linear time *without* any other variables?
+		<Quiz.Answer correct>
+			`void dfs(int v, int p, vector<vector<int> >& adj)`
+			<Quiz.Explanation>
+				Correct! For a tree, we do not need a visited array. We only need to check if we are revisiting the parent node, so `v` as a current node and `p` as a parent node would work.
+
+				Calling `adj` as a `vector<vector<int> >&` instead of just `vector<vector<int> >` ensures that the pointer is copied over each time, which takes constant instead of linear time. Therefore, the entire DFS would take linear time.
```
Comment (bqi343) [inline content/3_Silver/DFS.mdx:598]: reference isn't the same as pointer

File: content/3_Silver/DFS.mdx [hunk 10]
```diff
@@ -585,3 +585,77 @@ public class BuildingTeams
...
+		<Quiz.Answer>
+			`void dfs(int v, vector<vector<int> > adj)`
+			<Quiz.Explanation>
+				Incorrect. This DFS function cannot check for a parent node, and therefore would never terminate.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+		<Quiz.Answer>
+			`void dfs(vector<vector<int> > adj)`
+			<Quiz.Explanation>
+				Incorrect. This DFS function cannot track the current node, therefore it cannot even run.
```
Comment (bqi343) [inline content/3_Silver/DFS.mdx:616]: technically you could append to the adjacency list to keep track of the current state of the DFS? Probably better to use `const vector<vector<int>>& adj`

File: content/3_Silver/DFS.mdx [hunk 11]
```diff
@@ -585,3 +585,77 @@ public class BuildingTeams
...
+			</Quiz.Explanation>
+		</Quiz.Answer>
+		<Quiz.Answer>
+			Run DFS on the left node, add the current node to the list, and then run DFS on the right node.
+			<Quiz.Explanation>
+				Incorrect. This is inorder traversal.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+		<Quiz.Answer>
+			Run DFS on the right node, add the current node to the list, and then run DFS on the left node.
```
Comment (bqi343) [inline content/3_Silver/DFS.mdx:636]: no this isn't inorder traversal?

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/3_Silver/DFS.mdx (hunk 11 above)
Comment (xyzyzl) [reply content/3_Silver/DFS.mdx:636]: yeah i meant to say it's none of them

File: content/3_Silver/DFS.mdx (hunk 4 above)
Comment (xyzyzl) [reply content/3_Silver/DFS.mdx:594]: honestly it's just a preference thing

File: content/3_Silver/Flood_Fill.mdx (hunk 8 above)
Comment (xyzyzl) [reply content/3_Silver/Flood_Fill.mdx:791]: i could add "with a standard floodfill"

File: content/3_Silver/DFS.mdx (hunk 11 above)
Comment (nwatx) [reply content/3_Silver/DFS.mdx:636]: ?

File: content/3_Silver/DFS.mdx [hunk 12]
```diff
@@ -585,3 +585,49 @@ public class BuildingTeams
 ### Problems
 
 <Problems problems="bip" hideSuggestProblemButton />
+
+## Quiz
+<Quiz>
+	<Quiz.Question>
+		What would be a valid function header for DFS on an undirected tree to work in linear time *without* any other variables?
```
Comment (nwatx) [inline content/3_Silver/DFS.mdx:592]: should go in tree

**Review by Aadit-Ambadkar — APPROVED** [review]

File: content/3_Silver/DFS.mdx [hunk 13]
```diff
@@ -585,3 +585,35 @@ public class BuildingTeams
 ### Problems
 
 <Problems problems="bip" hideSuggestProblemButton />
+
+## Quiz
+<Quiz>
+	<Quiz.Question>
+		Can a DFS be used to find shortest paths in linear time?
```
Comment (nwatx) [inline content/3_Silver/DFS.mdx:592]: i mean you can if it's a linkedlist 😆

File: content/3_Silver/Flood_Fill.mdx [hunk 14]
```diff
@@ -739,3 +739,66 @@ print(room_num)
 ## Problems
 
 <Problems problems="general" hideSuggestProblemButton />
+
+## Quiz
+
+<Quiz>
+	<Quiz.Question>
+		What is the minimum number of starting nodes for flood fills we would need to visit every node in an undirected graph with four components?
```
Comment (nwatx) [inline content/3_Silver/Flood_Fill.mdx:747]:
```suggestion
		What is the minimum number of starting nodes we must "flood fill" to visit every node in an undirected graph with four components?
```

File: content/3_Silver/Flood_Fill.mdx [hunk 15]
```diff
@@ -739,3 +739,66 @@ print(room_num)
...
+		<Quiz.Answer correct>
+			4
+			<Quiz.Explanation>
+				Correct! Each DFS can travel one connected component, so a graph with four connected components would require **4** DFS runs.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+	</Quiz.Question>
+
+	<Quiz.Question>
+		Which of the following is not information we could get from the results of an unmodified floodfill algorithm?
```
Comment (nwatx) [inline content/3_Silver/Flood_Fill.mdx:775]: also add the header e.g. `ff(int x, int y)`

File: content/3_Silver/Flood_Fill.mdx [hunk 16]
```diff
@@ -739,3 +739,66 @@ print(room_num)
...
+		<Quiz.Answer>
+			The number of connected components of a graph
+			<Quiz.Explanation>
+				Incorrect. We find the number of connected components by tracking the number of times DFS is executed.
```
Comment (nwatx) [inline content/3_Silver/Flood_Fill.mdx:779]: technically it's executed a lot of times, you probably mean the amount of initial calls

File: content/3_Silver/Flood_Fill.mdx (hunk 15 above)
Comment (nwatx) [reply content/3_Silver/Flood_Fill.mdx:775]: also, isn't the number of connected components dependent on multiple flood fills and not "an unmodified flood fill algorithm."

**Review by nwatx — CHANGES_REQUESTED** [review]

**Review by dutinmeow — APPROVED** [review]

File: content/3_Silver/Flood_Fill.mdx [hunk 17]
```diff
@@ -739,3 +739,66 @@ print(room_num)
...
+		<Quiz.Answer correct>
+			4
+			<Quiz.Explanation>
+				Correct! Each DFS can travel one connected component, so a graph with four connected components would require **4** DFS runs.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+	</Quiz.Question>
+
+	<Quiz.Question>
+		Which of the following is not information we could get from the results of some number of iterations of a standard floodfill algorithm `void ff(int x, int y)`?
```
Comment (nwatx) [inline content/3_Silver/Flood_Fill.mdx:775]:
```suggestion
		Which of the following is **not** information we could get from the results of some number of iterations of a standard floodfill algorithm `void ff(int x, int y)`?
```

**Review by nwatx — APPROVED** [review]

File: content/3_Silver/DFS.mdx [hunk 18]
```diff
@@ -585,3 +585,35 @@ public class BuildingTeams
...
 <Problems problems="bip" hideSuggestProblemButton />
+
+## Quiz
+<Quiz>
+	<Quiz.Question>
+		Can a DFS with adjacency list be used to find shortest paths in linear time?
+		<Quiz.Answer>
+			Yes
+			<Quiz.Explanation>
+				Incorrect. Worst case, a DFS shortest path algorithm would have to iterate through <i>every</i> path before finding the shortest. Working shortest path algorithms will be covered later in the Gold section.
```
Comment (bqi343) [inline content/3_Silver/DFS.mdx:596]: No need to duplicate stuff between answers considering that all of them are shown.

File: content/3_Silver/DFS.mdx [hunk 19]
```diff
@@ -585,3 +585,35 @@ public class BuildingTeams
...
+			<Quiz.Explanation>
+				Correct! Worst case, a DFS shortest path algorithm would have to iterate through <i>every</i> path before finding the shortest. Working shortest path algorithms will be covered later in the Gold section.
+			</Quiz.Explanation>
+		</Quiz.Answer>
+	</Quiz.Question>
+
+	<Quiz.Question>
+		If a graph has two connected components, does there exist a node where we could start a DFS that visits every node in the graph?
+		<Quiz.Answer>
+			Yes
```
Comment (bqi343) [inline content/3_Silver/DFS.mdx:610]: Add an explanation here with "Incorrect." for consistency.

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Comment by SansPapyrus683** [comment]
image
we're all going to die

**Comment by bqi343** [comment]
Why does this PR now touch >1000 files? If you don't know how to resolve this, maybe create a new PR?

**Comment by dutinmeow** [comment]
wtf

**Comment by nwatx** [comment]
can you make a new PR or resolve this @xyzyzl

**Comment by xyzyzl** [comment]
what the fuck is going on

**Comment by xyzyzl** [comment]
yeah i'm making a new pull request

**Pull request closed without merging**